File: paleo_workbench/mapping/vector_lod.py
# -*- coding: utf-8 -*-
"""屏幕空间自适应矢量 LOD（vector-perf-increment Ticket 3）。

Visvalingam-Whyatt 有效面积化简的批量（向量化）实现：逐轮计算候选点
相对当前保留邻居的有效三角形面积，剔除低于容差的局部极小点，直至收
敛。地质形态特征点永不下桌（锚点）：

- 折线端点（环取首点；闭合点随环语义保留）；
- 拐点（相邻边叉积符号翻转处——蜿蜒岸线的折返特征）；
- 局部坐标极值（x/y 最小/最大——凸包意义上的形态支撑点）。

化简只作用于绘制管线（``_PreparedLayer`` 下游的显示几何），永不写回
要素几何——编辑拾取/拓扑始终在原始几何上进行（04-known-limitations
#4）。C++ 侧 ``map_edit_core.vector_lod_simplify`` 与本实现同运算序，
位级一致（tests/perf/test_vector_lod_bench.py 双实现互证）。

容差由视口比例推导：地图单位下的 ``pixel_tolerance × mupp`` 决定可剔
除三角形的有效面积上限（平方量纲）。
"""

from __future__ import annotations

import math

import numpy as np

#: 每顶点像素容差（0.75px：亚像素顶点对屏幕无贡献）。
DEFAULT_PIXEL_TOLERANCE = 0.75

#: 低于该顶点数的层不做数据集级化简（帧级像素量化已覆盖）。
MIN_VERTEX_COUNT = 4_000


def scale_bucket_mupp(mupp: float) -> float:
    """把 map-units-per-pixel 量化到 2 的幂档位（缓存键稳定性）。"""
    if mupp <= 0.0 or not math.isfinite(mupp):
        return 0.0
    return 2.0 ** math.floor(math.log2(mupp))


def tolerance_area(mupp: float,
                   pixel_tolerance: float = DEFAULT_PIXEL_TOLERANCE) -> float:
    """地图单位下的有效面积容差（三角形面积，平方量纲）。"""
    if mupp <= 0.0:
        return 0.0
    edge = pixel_tolerance * mupp
    return edge * edge


def _anchor_mask(
    xs: np.ndarray,
    ys: np.ndarray,
    starts: np.ndarray,
    ends: np.ndarray,
    is_ring: np.ndarray,
    tolerance: float,
) -> np.ndarray:
    """逐 part 标记形态锚点（端点 / 可见拐点 / 坐标极值）。

    拐点锚定是刻度相关的：噪声级微折（有效面积 < tolerance，本就不可
    见）不锚，否则带噪声的实测线会锁死全部顶点；只有 |cross| ≥ 2×容差
    （可见的方向改变）才锚——拐点保真与化简率在此取得一致。
    """
    n = len(xs)
    anchors = np.zeros(n, dtype=bool)
    for part in range(len(starts)):
        s, e = int(starts[part]), int(ends[part])
        if e - s < 3:
            anchors[s:e] = True
            continue
        px = xs[s:e]
        py = ys[s:e]
        anchors[s] = True
        anchors[e - 1] = True
        if is_ring[part]:
            # 闭合环：末点==首点，候选域 [s+1, e-1)。
            ax, ay = px[:-1], py[:-1]
            lo, hi = s + 1, e - 1
        else:
            ax, ay = px, py
            lo, hi = s + 1, e - 1
        local = anchors[lo:hi]
        if local.size >= 3:
            d1x = ax[1:-1] - ax[:-2]
            d1y = ay[1:-1] - ay[:-2]
            d2x = ax[2:] - ax[1:-1]
            d2y = ay[2:] - ay[1:-1]
            cross = d1x * d2y - d1y * d2x
            # turns[t] ⟷ 顶点 ax[t+2] ⟷ local[t+1]（local[0] ⟷ ax[1]）。
            # 仅可见拐点（|cross| ≥ 2×容差，即面积 ≥ 容差的转向）锚定。
            visible = np.abs(cross) >= 2.0 * tolerance
            turns = (cross[:-1] * cross[1:] < 0.0) & visible[:-1] & visible[1:]
            local[1:1 + turns.size] |= turns
            # 坐标极值（形态支撑点；平局取首现，与 C++ 一致）。
            # local[i] ⟷ ax[i+1]：argmin(arr[1:-1]) 的绝对 ax 下标为
            # m+1，对应 local 槽位 m。
            for arr in (ax, ay):
                local[int(np.argmin(arr[1:-1]))] = True
                local[int(np.argmax(arr[1:-1]))] = True
    return anchors


def visvalingam_keep_mask(
    xs: np.ndarray,
    ys: np.ndarray,
    starts: np.ndarray,
    is_ring: np.ndarray,
    tolerance: float,
) -> np.ndarray:
    """批量 Visvalingam-Whyatt：返回等长保留掩码（True=保留顶点）。

    拼接布局：``xs/ys`` 全部顶点；``starts`` 各 part 起始下标（末 part
    到 ``len(xs)`` 结束）；``is_ring`` 闭合环标记。``tolerance <= 0``
    恒等（全保留）。

    逐轮：非锚内部点以其当前保留邻居计算有效三角形面积；低于容差的
    连续带只剔除局部极小（同轮互为邻居的剔除会破坏面积语义），直至
    无可剔除——剩余候选全部保留。邻居关系经保留掩码的前/后向累积索
    引重建；part 边界是锚点，链天然不跨界。
    """
    n = len(xs)
    if n == 0 or tolerance <= 0.0:
        return np.ones(n, dtype=bool)
    ends = np.append(starts[1:], n)
    anchors = _anchor_mask(xs, ys, starts, ends, is_ring, tolerance)
    # 经典 VW 语义：三角形面积相对“当前存活邻居”链评估——keep 从全
    # True 起步随剔除收缩（若从锚点起步，首轮邻居即远距锚点，三角形
    # 面积巨大，什么都剔不掉）。
    keep = np.ones(n, dtype=bool)
    pending = np.ones(n, dtype=bool)
    pending &= ~anchors  # 候选 = 非锚内部点（锚点已定为保留）
    idx = np.arange(n)
    part_of = np.searchsorted(starts, idx, side="right") - 1
    interior_pos = (idx > starts[part_of]) & (idx < ends[part_of] - 1)
    pending &= interior_pos

    while True:
        if not pending.any():
            break
        # 严格 j<i / j>i 的最近保留邻居（含自身的退化三角形面积为 0）。
        accum = np.maximum.accumulate(np.where(keep, idx, -1))
        prev_kept = np.concatenate(([-1], accum[:-1]))
        rev_min = np.minimum.accumulate(
            np.where(keep[::-1], idx[::-1], n).reshape(-1))[::-1]
        nxt = np.concatenate((rev_min[1:], [n]))
        has_both = (prev_kept >= 0) & (nxt < n)
        pv = np.where(has_both, prev_kept, idx)
        nx = np.where(has_both, nxt, idx)
        ax_ = xs - xs[pv]
        ay_ = ys - ys[pv]
        bx_ = xs[nx] - xs[pv]
        by_ = ys[nx] - ys[pv]
        areas = np.abs(ax_ * by_ - ay_ * bx_) * 0.5
        below = pending & has_both & (areas < tolerance)
        if not below.any():
            keep |= pending  # 无可剔除：剩余候选全部保留
            break
        remove = below.copy()
        rolled_left = np.roll(areas, 1)
        rolled_right = np.roll(areas, -1)
        left_below = np.zeros(n, dtype=bool)
        right_below = np.zeros(n, dtype=bool)
        left_below[1:] = below[:-1]
        right_below[:-1] = below[1:]
        # 连续低于带只剔除严格局部极小：对比项取 >=（非极小被排除）。
        remove &= ~(left_below & (areas >= rolled_left))
        remove &= ~(right_below & (areas >= rolled_right))
        if not remove.any():
            remove = below & ~(left_below | right_below)
        if not remove.any():
            # 整带同面积平局：每条连续 below 带内隔点剔除（每轮 O(带数)
            # 进度，避免「全局只剔 1 点」在万级顶点上退化成 O(n²) 帧挂死
            # ——CI render threaded/cancel 与大图层交互的根因）。
            below_idx = np.nonzero(below)[0]
            remove = np.zeros(n, dtype=bool)
            if below_idx.size:
                run_starts = np.empty(below_idx.size, dtype=bool)
                run_starts[0] = True
                if below_idx.size > 1:
                    run_starts[1:] = np.diff(below_idx) > 1
                # 带内位置：相对本带起点的偏移。
                start_pos = np.maximum.accumulate(
                    np.where(run_starts, np.arange(below_idx.size), 0))
                pos_in_run = np.arange(below_idx.size) - start_pos
                remove[below_idx[pos_in_run % 2 == 0]] = True
        # remove 本轮剔除：落 keep 位 + 退出候选；带内非极小点留待下轮
        # 以新邻居重算——仍候选。
        keep &= ~remove
        pending &= ~remove
    return keep

File: libs/ui_canvas/src/vector_lod.cpp
#include <pwb/ui_canvas/vector_lod.hpp>

#include <algorithm>
#include <cmath>
#include <limits>
#include <numeric>

namespace pwb::ui_canvas::vector_lod {

double scale_bucket_mupp(double mupp) {
    if (mupp <= 0.0 || !std::isfinite(mupp)) return 0.0;
    return std::pow(2.0, std::floor(std::log2(mupp)));
}

double tolerance_area(double mupp, double pixel_tolerance) {
    if (mupp <= 0.0) return 0.0;
    const double edge = pixel_tolerance * mupp;
    return edge * edge;
}

namespace {

// _anchor_mask — per-part morphology anchors: endpoints, VISIBLE
// inflections (cross-sign flips where both adjacent turns have
// |cross| >= 2*tolerance), and x/y extrema (first occurrence on ties).
std::vector<bool> anchor_mask(const std::vector<double>& xs,
                              const std::vector<double>& ys,
                              const std::vector<std::size_t>& starts,
                              const std::vector<std::size_t>& ends,
                              const std::vector<bool>& is_ring,
                              double tolerance) {
    const std::size_t n = xs.size();
    std::vector<bool> anchors(n, false);
    for (std::size_t part = 0; part < starts.size(); ++part) {
        const std::size_t s = starts[part];
        const std::size_t e = ends[part];
        if (e - s < 3) {
            for (std::size_t i = s; i < e; ++i) anchors[i] = true;
            continue;
        }
        anchors[s] = true;
        anchors[e - 1] = true;
        // Closed ring: last vertex duplicates the first — the candidate
        // domain drops it (ax = px[:-1]); open parts keep all vertices.
        const std::size_t ax_len = is_ring[part] ? (e - s - 1) : (e - s);
        // local[i] <-> anchors[s+1+i] <-> ax[i+1].
        const std::size_t local_size = e - s - 2;
        if (local_size >= 3) {
            const double* ax = xs.data() + s;
            const double* ay = ys.data() + s;
            // cross[t] = turn at ax[t+1], t in 0..ax_len-3.
            if (ax_len >= 3) {
                std::vector<double> cross(ax_len - 2);
                for (std::size_t t = 0; t + 2 < ax_len; ++t) {
                    const double d1x = ax[t + 1] - ax[t];
                    const double d1y = ay[t + 1] - ay[t];
                    const double d2x = ax[t + 2] - ax[t + 1];
                    const double d2y = ay[t + 2] - ay[t + 1];
                    cross[t] = d1x * d2y - d1y * d2x;
                }
                const double threshold = 2.0 * tolerance;
                // turns[t] <-> ax[t+2] <-> local[t+1].
                for (std::size_t t = 0; t + 1 < cross.size(); ++t) {
                    const bool visible0 =
                        std::abs(cross[t]) >= threshold;
                    const bool visible1 =
                        std::abs(cross[t + 1]) >= threshold;
                    if (cross[t] * cross[t + 1] < 0.0 && visible0 &&
                        visible1) {
                        anchors[s + 1 + (t + 1)] = true;
                    }
                }
                // Coordinate extrema of arr[1:-1] — first occurrence on
                // ties (np.argmin/argmax parity).
                for (const double* arr : {ax, ay}) {
                    if (ax_len >= 2) {
                        std::size_t lo = 0, hi = 0;
                        for (std::size_t i = 1; i + 1 < ax_len; ++i) {
                            if (arr[i] < arr[lo + 1]) lo = i - 1;
                            if (arr[i] > arr[hi + 1]) hi = i - 1;
                        }
                        // lo/hi are indices into arr[1:-1]; map back:
                        // arr[1:-1][m] <-> ax[m+1] <-> local[m].
                        anchors[s + 1 + lo] = true;
                        anchors[s + 1 + hi] = true;
                    }
                }
            }
        }
    }
    return anchors;
}

}  // namespace

std::vector<bool> visvalingam_keep_mask(
    const std::vector<double>& xs, const std::vector<double>& ys,
    const std::vector<std::size_t>& starts,
    const std::vector<bool>& is_ring, double tolerance) {
    const std::size_t n = xs.size();
    if (n == 0 || tolerance <= 0.0) {
        return std::vector<bool>(n, true);
    }
    std::vector<std::size_t> ends(starts.begin() + 1, starts.end());
    ends.push_back(n);
    const std::vector<bool> anchors =
        anchor_mask(xs, ys, starts, ends, is_ring, tolerance);

    std::vector<bool> keep(n, true);
    // pending = ~anchors & interior (part boundaries are anchors anyway,
    // but the Python mask applies interior_pos explicitly).
    std::vector<bool> pending(n, false);
    std::vector<std::size_t> part_of(n, 0);
    {
        std::size_t part = 0;
        for (std::size_t i = 0; i < n; ++i) {
            while (part + 1 < starts.size() && starts[part + 1] <= i) {
                ++part;
            }
            part_of[i] = part;
            const bool interior =
                i > starts[part] && i + 1 < ends[part];
            pending[i] = !anchors[i] && interior;
        }
    }

    std::vector<std::size_t> idx(n);
    std::iota(idx.begin(), idx.end(), 0);
    std::vector<std::size_t> prev_kept(n), nxt(n);
    std::vector<double> areas(n);
    std::vector<bool> below(n), remove(n), left_below(n), right_below(n);

    for (;;) {
        bool any_pending = false;
        for (std::size_t i = 0; i < n; ++i) any_pending |= pending[i];
        if (!any_pending) break;

        // prev_kept[i] = greatest kept index < i (nearest surviving
        // neighbour left); nxt[i] = smallest kept index > i.
        {
            // run_max = maximum.accumulate(where(keep, idx, -1));
            // prev_kept shifts it right with a -1 front.
            std::vector<std::size_t> run_max(n, 0);
            std::vector<bool> run_have(n, false);
            std::size_t accum = 0;
            bool have = false;
            for (std::size_t i = 0; i < n; ++i) {
                if (keep[i] && (!have || i > accum)) {
                    accum = i;
                    have = true;
                }
                run_max[i] = accum;
                run_have[i] = have;
            }
            for (std::size_t i = 0; i < n; ++i) {
                prev_kept[i] = (i > 0 && run_have[i - 1])
                                   ? run_max[i - 1]
                                   : std::numeric_limits<std::size_t>::max();
            }
            // nxt via reverse running min of kept idx, shifted left.
            std::vector<std::size_t> run_min(n, n);
            std::size_t rmin = n;
            for (std::size_t i = n; i-- > 0;) {
                if (keep[i] && i < rmin) rmin = i;
                run_min[i] = rmin;
            }
            for (std::size_t i = 0; i < n; ++i) {
                nxt[i] = (i + 1 < n) ? run_min[i + 1] : n;
            }
        }
        const std::size_t NPOS = std::numeric_limits<std::size_t>::max();
        std::vector<bool> has_both(n);
        for (std::size_t i = 0; i < n; ++i) {
            has_both[i] = (prev_kept[i] != NPOS) && (nxt[i] < n);
        }
        for (std::size_t i = 0; i < n; ++i) {
            const std::size_t pv = has_both[i] ? prev_kept[i] : i;
            const std::size_t nx = has_both[i] ? nxt[i] : i;
            const double ax_ = xs[i] - xs[pv];
            const double ay_ = ys[i] - ys[pv];
            const double bx_ = xs[nx] - xs[pv];
            const double by_ = ys[nx] - ys[pv];
            areas[i] = std::abs(ax_ * by_ - ay_ * bx_) * 0.5;
            below[i] = pending[i] && has_both[i] && areas[i] < tolerance;
        }
        bool any_below = false;
        for (std::size_t i = 0; i < n; ++i) any_below |= below[i];
        if (!any_below) {
            // Nothing removable: all remaining candidates keep.
            for (std::size_t i = 0; i < n; ++i) {
                if (pending[i]) keep[i] = true;
            }
            break;
        }
        // np.roll wraparound: rolled_left[i] = areas[(i-1+n)%n].
        std::vector<double> rolled_left(n), rolled_right(n);
        for (std::size_t i = 0; i < n; ++i) {
            rolled_left[i] = areas[(i + n - 1) % n];
            rolled_right[i] = areas[(i + 1) % n];
        }
        for (std::size_t i = 0; i < n; ++i) {
            left_below[i] = i > 0 ? below[i - 1] : false;
            right_below[i] = i + 1 < n ? below[i + 1] : false;
        }
        // Consecutive-below bands drop only strict local minima (the
        // comparisons take >= so non-minima survive the round).
        for (std::size_t i = 0; i < n; ++i) {
            remove[i] = below[i] &&
                        !(left_below[i] && areas[i] >= rolled_left[i]) &&
                        !(right_below[i] && areas[i] >= rolled_right[i]);
        }
        bool any_remove = false;
        for (std::size_t i = 0; i < n; ++i) any_remove |= remove[i];
        if (!any_remove) {
            for (std::size_t i = 0; i < n; ++i) {
                remove[i] = below[i] && !(left_below[i] || right_below[i]);
            }
        }
        any_remove = false;
        for (std::size_t i = 0; i < n; ++i) any_remove |= remove[i];
        if (!any_remove) {
            // Equal-area tie band: thin every other vertex inside each
            // consecutive below-run (O(#runs) progress per round).
            std::vector<std::size_t> below_idx;
            for (std::size_t i = 0; i < n; ++i) {
                if (below[i]) below_idx.push_back(i);
            }
            for (std::size_t k = 0; k < below_idx.size(); ++k) {
                // pos_in_run: offset from this run's start.
                std::size_t run_start = 0;
                for (std::size_t j = k; j > 0; --j) {
                    if (below_idx[j] - below_idx[j - 1] > 1) {
                        run_start = j;
                        break;
                    }
                }
                const std::size_t pos = k - run_start;
                remove[below_idx[k]] = (pos % 2 == 0);
            }
        }
        for (std::size_t i = 0; i < n; ++i) {
            if (remove[i]) {
                keep[i] = false;
                pending[i] = false;
            }
        }
    }
    return keep;
}

}  // namespace pwb::ui_canvas::vector_lod
